packages/framework-types/src/concepts/token-verifier.ts

```diff
@@ -0,0 +1,27 @@
+import { UserEnvelope } from '../envelope'
+
+interface ITokenVerifier {
+  readonly issuer: string
+  readonly rolesClaim: string
+  readonly extraValidation?: (jwtToken: Record<string, unknown>, rawToken: string) => Promise<void>
+
+  verify?(token: string): Promise<UserEnvelope>
+}
+
+export interface JwskUriTokenVerifier extends ITokenVerifier {
+  readonly jwksUri: string
+}
+
+export interface PublicKeyTokenVerifier extends ITokenVerifier {
+  readonly publicKey: Promise<string>
+}
+
+export type TokenVerifier = JwskUriTokenVerifier | PublicKeyTokenVerifier
```

Comment:
I'd avoid relying only on inheritance to avoid explicit casts/checks by callers (like the `has*` functions defined below).

A tagged union would work well here :)

```suggestion
interface TokenVerifierCommon {
  readonly issuer: string
  readonly rolesClaim: string
  readonly extraValidation?: (jwtToken: Record<string, unknown>, rawToken: string) => Promise<void>
  verify?(token: string): Promise<UserEnvelope>
}

type TokenVerifierInfo =
  | { readonly _tag: 'JwksUriTokenVerifier', readonly jwksUri: string }
  | { readonly _tag: 'PublicKeyTokenVerifier', readonly publicKey: Promise<string> }

export type TokenVerifier = TokenVerifierCommon & TokenVerifierInfo
```

This way, if someone does `if (tokenVerifier._tag === '...')`, not only they will get autocompletion in their IDE for the `_tag`, but also inside of the `if` TS will know the appropriate available info fields